onnx/version_converter/convert.h

```diff
@@ -0,0 +1,160 @@
+        if (it->domain() == initial_version.domain()) {
+          ONNX_ASSERTM(initial_version.version() == it->version(),
+              "initial_version does not reflect current state of model");
+        }
+      }
+
+      std::shared_ptr<Graph> g(ImportModelProto(mp_in));
+      ONNX_ASSERTM(g.get() != nullptr,
+        "Warning: onnx version converter is unable to parse input model. "
+        "(The IR version of the ONNX model may be too old.)");
+
+      // TODO: Move to Inter-Domain Converter
+      // Get initial model versions
+      // std::vector<OpSetID> initial_versions = g->opset_versions_mutable();
+
+      // No conversion necessary if Model has single, equivalent opset version
+      // if (initial_versions.size() == 1 && initial_versions[0].version ==
+      //    target_version.version && initial_versions[0].domain ==
+      //    target_version.domain) {
+      //  return mp_in;
+      // }
+
+      // Check if target_version is valid
+      const std::unordered_map<std::string, std::pair<int, int>>& versions_map = OpSchemaRegistry::DomainToVersionRange::Instance().Map();
+      const std::string search_domain = target_version.domain() == "ai.onnx" ? "" : target_version.domain();
+      const std::pair<int, int>& version_range = versions_map.at(search_domain);
+      ONNX_ASSERTM(target_version.version() >= version_range.first && target_version
+          .version() <= version_range.second,
+        "Warning: invalid target_version (must be between %s and %s",
+        version_range.first, version_range.second);
+      // Compile list of all ops used in the model
+      graph_node_list nodes = g->nodes();
+
+      const std::vector<OpSchema>& all_opschemas = OpSchemaRegistry::get_all_schemas_with_history();
+
+      // Create Map for All Versions of format {op_name: {domain: {version: schema}}}
+      std::unordered_map<std::string, std::unordered_map<std::string, std::map<int64_t, const OpSchema*>>>  all_schemas;
+
+      for (const OpSchema& schema : all_opschemas) {
+        all_schemas[schema.Name()][schema.domain()][(int64_t) schema.since_version()] = &schema;
+      }
+
+      // Create Map for Current Version
+      std::unordered_map<const Node*, const OpSchema*> current_opschemas;
+      for (const Node* op : nodes) {
+        // Iterate through all OperatorSetVersions, select highest that is leq initial_version
+        int64_t op_opset_version = -1;
+        auto& op_domain_map = all_schemas[op->kind().toString()];
+        if (op_domain_map.find(initial_domain) != op_domain_map.end()) {
+          // If op isn't defined for initial domain, we won't convert it
+          for (const auto& version_pair : op_domain_map[initial_domain]) {
+            if (version_pair.first > op_opset_version && version_pair.first <= initial_version.version()) {
+              op_opset_version = version_pair.first;
+              current_opschemas[op] = op_domain_map.at(initial_domain).at(op_opset_version);
```

Comment:
This current_opschemas seems not quite useful.

Reply:
The idea is just to store the current version of each op, which was useful before I switched to using since_version.  I will attempt to phase it out.